org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java

```diff
@@ -198,16 +320,18 @@ public boolean isCanceled() {
 				};
 			}
 			SubMonitor progress = SubMonitor.convert(monitor, 1);
-			if (kind == FULL_BUILD) {
-				fullBuild(progress.split(1), IBuildFlag.RECOVERY_BUILD.isSet(args));
+			if (kind == FULL_BUILD || wasFullBuildRequested) {
+				fullBuild(progress.split(1), !wasFullBuildRequested && IBuildFlag.RECOVERY_BUILD.isSet(args));
 			} else {
 				IResourceDelta delta = getDelta(getProject());
 				if (delta == null || isOpened(delta)) {
-					fullBuild(progress.split(1), IBuildFlag.RECOVERY_BUILD.isSet(args));
+					fullBuild(progress.split(1), !wasFullBuildRequested && IBuildFlag.RECOVERY_BUILD.isSet(args));
 				} else {
 					incrementalBuild(delta, progress.split(1));
 				}
 			}
+			// If no exceptions have been thrown, fix the information about the requested full builds
```

Comment:
Just trying to clarify the scenario.

IIUC, we would call `requestFullBuild`. This will trigger an autobuild. The autobuild will then somewhere along the line call this method.

In this case we have `nextBuildIsFullBuild` be true here and `allowRequestFullBuild` false. If an exception did now occur, we couldn't request a new full build, the next build will be a full build, but there currently (possibly) is no more build.

In short, if an exception occurred, it's possible that we have an unbuilt project that we do not retry to build.

Reply:
The AutoBuildJob will be rescheduled in case of cancellation.